deepinv/physics/generator/base.py

```diff
@@ -26,41 +29,81 @@ class PhysicsGenerator(nn.Module):
-        >>> _ = torch.manual_seed(0)
         >>> # combine a PhysicsGenerator for blur and noise level parameters
         >>> generator = MotionBlurGenerator(psf_size = (3, 3), num_channels = 1) + SigmaGenerator()
-        >>> params_dict = generator.step(batch_size=1) # dict_keys(['filter', 'sigma'])
+        >>> params_dict = generator.step(batch_size=1, seed=0) # dict_keys(['filter', 'sigma'])
         >>> print(params_dict['filter'])
         tensor([[[[0.0000, 0.1006, 0.0000],
                   [0.0000, 0.8994, 0.0000],
                   [0.0000, 0.0000, 0.0000]]]])
         >>> print(params_dict['sigma'])
-        tensor([0.1577])
+        tensor([0.2532])
     """
 
     def __init__(
-        self, step=lambda **kwargs: {}, device="cpu", dtype=torch.float32, **kwargs
+        self,
+        step=lambda **kwargs: {},
+        rng: torch.Generator = None,
+        device="cpu",
+        dtype=torch.float32,
+        **kwargs,
     ) -> None:
         super().__init__()
 
         self.step_func = step
         self.kwargs = kwargs
         self.factory_kwargs = {"device": device, "dtype": dtype}
+        self.device = device
+        if rng is None:
+            self.rng = torch.Generator(device=device)
+        else:
+            # Make sure that the random generator is on the same device as the physic generator
+            assert (
+                rng.device == device
+            ), "The random generator is not on the same device as the Physic Generator"
+            self.rng = rng
+        self.initial_random_state = self.rng.get_state()
+
         # Set attributes
         for k, v in kwargs.items():
             setattr(self, k, v)
 
-    def step(self, batch_size=1, **kwargs):
+    def step(self, batch_size: int = 1, seed: int = None, **kwargs):
         r"""
         Generates a batch of parameters for the forward operator.
 
         :param int batch_size: the number of samples to generate.
+        :param int seed: the seed for the random number generator.
         :returns: A dictionary with the new parameters, that is ``{param_name: param_value}``.
         """
-        if not kwargs:
+        self.rng_manual_seed(seed)
+        if kwargs is not None:
             self.kwargs = kwargs
 
-        return self.step_func(**kwargs)
+        return self.step_func(batch_size, seed, **kwargs)
+
+    def rng_manual_seed(self, seed: int = None):
+        r"""
+        Sets the seed for the random number generator.
+
+        :param int seed: the seed to set for the random number generator.
+         If not provided, the current state of the random number generator is used.
+         Note: The `torch.manual_seed` is triggered when a the random number generator is not initialized.
+        """
+        if seed is not None:
+            if self.rng is not None:
+                self.rng = self.rng.manual_seed(seed)
+            else:
+                torch.manual_seed(seed)
```

Comment:
Currently all randomness in our library come from torch. torch.manual_seed() will set the random seed for all devices. Your function is good for seed everything but It might cause some problems for other randomness? Also, in the __init__, if a rng is not provided, a new rng on the same device will be created under the hood so the call to torch.manual_seed() might disappear.

Reply:
I agree with @mh-nguyen712 - I think our code should remain functional and not interfere with the global state - if the user really wants to set a universal seed maybe we can add the function that Matthieu suggested in `deepinv.utils` as a utility function?